src/base/static/state/ducks/map.js

```diff
@@ -1,311 +1,581 @@
       return {
         ...state,
-        layerStatuses: {
-          ...state.layerStatuses,
-          ...newStatuses,
+        style: {
+          ...state.style,
+          sources: {
+            ...state.style.sources,
+            "__mapseed-focused-source__": {
+              type: "geojson",
+              data: {
+                type: "FeatureCollection",
+                features: action.payload,
+              },
+            },
+          },
         },
       };
-    case UPDATE_LAYERS_VISIBLE:
-      newStatuses = action.payload.reduce((memo, layerId) => {
-        return {
-          ...memo,
-          [layerId]: {
-            ...state.layerStatuses[layerId],
-            isVisible: true,
-          },
-        };
-      }, {});
-
+    case UPDATE_VIEWPORT:
       return {
         ...state,
-        layerStatuses: {
-          ...state.layerStatuses,
-          ...newStatuses,
+        viewport: {
+          ...state.viewport,
+          ...action.payload.viewport,
+          // NOTE: This is a fix for an apparent bug in react-map-gl.
+          // See: https://github.com/uber/react-map-gl/issues/630
+          bearing: isNaN(action.payload.viewport.bearing)
+            ? state.viewport.bearing
+            : action.payload.viewport.bearing,
+          // These checks support a "scroll zoom around center" feature (in
+          // which a zoom of the map will not change the centerpoint) that is
+          // not exposed by react-map-gl. These checks are pretty convoluted,
+          // though, so it would be great if react-map-gl could just
+          // incorporate the scroll zoom around center option natively.
+          // See: https://github.com/uber/react-map-gl/issues/515
+          latitude:
+            action.payload.scrollZoomAroundCenter &&
+            action.payload.viewport.zoom !== state.viewport.zoom
+              ? state.viewport.latitude
+              : action.payload.viewport.latitude
+                ? action.payload.viewport.latitude
+                : state.viewport.latitude,
+          longitude:
+            action.payload.scrollZoomAroundCenter &&
+            action.payload.viewport.zoom !== state.viewport.zoom
+              ? state.viewport.longitude
+              : action.payload.viewport.longitude
+                ? action.payload.viewport.longitude
+                : state.viewport.longitude,
         },
       };
-    case UPDATE_LAYERS_HIDDEN:
-      newStatuses = action.payload.reduce((memo, layerId) => {
-        return {
-          ...memo,
-          [layerId]: {
-            ...state.layerStatuses[layerId],
-            isVisible: false,
-          },
-        };
-      }, {});
-
+    case LOAD_STYLE_AND_METADATA:
       return {
         ...state,
-        layerStatuses: {
-          ...state.layerStatuses,
-          ...newStatuses,
+        style: {
+          ...state.style,
+          ...action.payload.style,
+        },
+        layerGroupsMetadata: {
```

Comment:
The `layerGroupsMetadata` data structure keeps track of overall layer group visibility, basemap status, and the ids of the actual mapbox layers that make up each layer group.

Reply:
Maybe we can add this as a code comment?